src/org/stringtemplate/v4/STGroup.java

```diff
@@ -745,8 +745,9 @@ public void registerModelAdaptor(Class<?> attributeType, ModelAdaptor adaptor) {
         adaptors.put(attributeType, adaptor);
     }
 
-    public ModelAdaptor getModelAdaptor(Class<?> attributeType) {
-        return adaptors.get(attributeType);
+    public <T> ModelAdaptor<T> getModelAdaptor(Class<? extends T> attributeType) {
```

Comment:
```suggestion
    public <T> ModelAdaptor<? super T> getModelAdaptor(Class<T> attributeType) {
```

Reply:
I did some more research and found that this is actually not the right way to use bounded wildcards. While `<T> ModelAdaptor<T> getModelAdaptor(Class<T>)` is definitely wrong because it does not account for subtypes, both `<T> ModelAdaptor<T> getModelAdaptor(Class<? extends T>)` and `<T> ModelAdaptor<? super T> getModelAdaptor(Class<T> attributeType)` correctly model the relationship between types. However, the first option is actually preferred:

https://stackoverflow.com/questions/22815023/generic-wildcard-types-should-not-be-used-in-return-parameters
https://stackoverflow.com/questions/59544179/what-is-the-more-idiomatic-use-of-super-and-extends-in-this-case (mine)

Reply:
The backing map (`TypeRegistry<T>`) does not behave in the manner you are probably thinking. The signature `<T> ModelAdaptor<? super T> getModelAdaptor(Class<T> attributeType)` is the only signature that accurately reflects the semantics of this type. This is observable in the following scenarios:

```
st.registerModelAdaptor(Exception.class, new ObjectModelAdaptor<Exception>());

// ✔ This line is correct:
ModelAdaptor<? super Exception> adaptor1 = st.<Exception>getModelAdaptor(new Exception().getClass());

// ❌ This line is not semantically correct:
ModelAdaptor<? super Object> adaptor2 = st.<Object>getModelAdaptor(new Exception().getClass());
```

Reply:
Am I not correct in that `TypeRegistry` behaves in the way that when I do `put(Object.class, foo)`, I can `get(String.class) == foo`, regardless of what the value type is?

Also, please elaborate on the scenario. Does it assume `<T> ModelAdaptor<? super T> getModelAdaptor(Class<T>)`? In that case both lines to not compile, because `getClass` has return type `Class<? extends <ReceiverType>>`:

![Screenshot 2020-01-01 at 16 41 56](https://user-images.githubusercontent.com/4145923/71643141-aceae780-2cb5-11ea-96fb-53cfc464243f.png)

With `Class<? extends T>` it compiles.

Indeed, this was already a problem when I added `Interpreter.renderObject` back in the AttributeRenderer PR. It's declaration is

```java
<T> String renderObject(InstanceScope scope, String formatString, T o, Class<? extends T> attributeType)
```

Without `? extends` the call `renderObject(scope, format, obj, obj.getClass())` would be invalid, and without `? extends` in `STGroup.getAttributeRenderer` the call `group.getRenderer(attributeType)` would not be valid.

Reply:
> Also, please elaborate on the scenario. Does it assume ... ?

It doesn't make particular assumptions about the signature. The primary point is the type of the return value when `getModelAdaptor(Exception.class)` is called.

It will not be possible to make the call sites correct without some sort of casting due to dynamic typing in StringTemplate. However, this limitation should not affect the signatures of methods where we do know the correct types.

Reply:
Ok, that sounds reasonable. Thanks!